### ISSUE TITLE
Failed to install rJava from CRAN in conda env on Linux

### ISSUE THREAD

**Issue by rolweber** [s-u/rJava#274]
Hello rJava team, I'd appreciate some help with debugging an installation problem.

I'm trying to install rJava from CRAN into a conda environment. I'm deliberately not using the r-rjava conda package, because that would pull in an OpenJDK packaged by Anaconda. I have to use the OpenJDK from the Linux distribution (RedHat UBI 8), which gets more frequent security updates.

The conda environment has r-base and a bunch of other packages installed. JAVA_HOME is set. I can activate the environment and run `R CMD javareconf` successfully.
```bash
source activate R-3.6-py38
R CMD javareconf
```
<details><pre>Java interpreter : /etc/alternatives/java_sdk_openjdk/jre/bin/java
Java version     : 1.8.0_302
Java home path   : /etc/alternatives/java_sdk_openjdk
Java compiler    : /etc/alternatives/java_sdk_openjdk/bin/javac
Java headers gen.: /etc/alternatives/java_sdk_openjdk/bin/javah
Java archive tool: /etc/alternatives/java_sdk_openjdk/bin/jar

trying to compile and link a JNI program 
detected JNI cpp flags    : -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/linux
detected JNI linker flags : -L/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server -ljvm
x86_64-conda_cos6-linux-gnu-cc -I"/opt/conda/envs/R-3.6-py38/lib/R/include" -DNDEBUG -I/etc/alternatives/java_sdk_openjdk/include -I/etc/alternatives/java_sdk_openjdk/include/linux  -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/R-3.6-py38/include -I/opt/conda/envs/R-3.6-py38/include -Wl,-rpath-link,/opt/conda/envs/R-3.6-py38/lib  -fpic  -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/R-3.6-py38/include -fdebug-prefix-map=/tmp/build/80754af9/r-base_1589917437985/work=/usr/local/src/conda/r-base-3.6.1 -fdebug-prefix-map=/opt/conda/envs/R-3.6-py38=/usr/local/src/conda-prefix  -c conftest.c -o conftest.o
x86_64-conda_cos6-linux-gnu-cc -shared -L/opt/conda/envs/R-3.6-py38/lib/R/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/R-3.6-py38/lib -Wl,-rpath-link,/opt/conda/envs/R-3.6-py38/lib -L/opt/conda/envs/R-3.6-py38/lib -L/opt/conda/envs/R-3.6-py38/lib -Wl,-rpath-link,/opt/conda/envs/R-3.6-py38/lib -o conftest.so conftest.o -L/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server -ljvm -L/opt/conda/envs/R-3.6-py38/lib/R/lib -lR

JAVA_HOME        : /etc/alternatives/java_sdk_openjdk
Java library path: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server
JNI cpp flags    : -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/linux
JNI linker flags : -L/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server -ljvm
Updating Java configuration in /opt/conda/envs/R-3.6-py38/lib/R
Done.
</pre></details>

Then when I try to install rJava using the recommended `install.packages('rJava')`, I'm getting an error:
```txt
R --no-save
install.packages('rJava')
```
```txt
checking Java support in R... present:
interpreter : '/etc/alternatives/java_sdk_openjdk/jre/bin/java'
archiver    : '/etc/alternatives/java_sdk_openjdk/bin/jar'
compiler    : '/etc/alternatives/java_sdk_openjdk/bin/javac'
header prep.: '/etc/alternatives/java_sdk_openjdk/bin/javah'
cpp flags   : '-I/etc/alternatives/java_sdk_openjdk/include -I/etc/alternatives/java_sdk_openjdk/include/linux'
java libs   : '-L/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server -ljvm'
checking whether Java run-time works... yes
checking whether -Xrs is supported... yes
checking whether -Xrs will be used... yes
checking whether JVM will be loaded dynamically... no
checking whether JNI programs can be compiled... configure: error: Cannot compile a simple JNI program. See config.log for details.

Make sure you have Java Development Kit installed and correctly registered in R.
If in doubt, re-run "R CMD javareconf" as root.

ERROR: configuration failed for package ‘rJava’
* removing ‘/opt/conda/envs/R-3.6-py38/lib/R/library/rJava’
```
See the details section for full output.
<details><pre>trying URL 'https://cran.uni-muenster.de/src/contrib/rJava_1.0-4.tar.gz'
Content type 'application/x-gzip' length 672236 bytes (656 KB)
==================================================
downloaded 656 KB

* installing *source* package ‘rJava’ ...
** package ‘rJava’ successfully unpacked and MD5 sums checked
** using staged installation
checking for gcc... x86_64-conda_cos6-linux-gnu-cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether x86_64-conda_cos6-linux-gnu-cc accepts -g... yes
checking for x86_64-conda_cos6-linux-gnu-cc option to accept ISO C89... none needed
checking how to run the C preprocessor... /opt/conda/envs/R-3.6-py38/bin/x86_64-conda_cos6-linux-gnu-cpp
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/wait.h that is POSIX.1 compatible... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for string.h... (cached) yes
checking sys/time.h usability... yes
checking sys/time.h presence... yes
checking for sys/time.h... yes
checking for unistd.h... (cached) yes
checking for an ANSI C-conforming const... yes
checking whether time.h and sys/time.h may both be included... yes
configure: checking whether x86_64-conda_cos6-linux-gnu-cc supports static inline...
yes
checking whether setjmp.h is POSIX.1 compatible... yes
checking whether sigsetjmp is declared... yes
checking whether siglongjmp is declared... yes
checking Java support in R... present:
interpreter : '/etc/alternatives/java_sdk_openjdk/jre/bin/java'
archiver    : '/etc/alternatives/java_sdk_openjdk/bin/jar'
compiler    : '/etc/alternatives/java_sdk_openjdk/bin/javac'
header prep.: '/etc/alternatives/java_sdk_openjdk/bin/javah'
cpp flags   : '-I/etc/alternatives/java_sdk_openjdk/include -I/etc/alternatives/java_sdk_openjdk/include/linux'
java libs   : '-L/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server -ljvm'
checking whether Java run-time works... yes
checking whether -Xrs is supported... yes
checking whether -Xrs will be used... yes
checking whether JVM will be loaded dynamically... no
checking whether JNI programs can be compiled... configure: error: Cannot compile a simple JNI program. See config.log for details.

Make sure you have Java Development Kit installed and correctly registered in R.
If in doubt, re-run "R CMD javareconf" as root.

ERROR: configuration failed for package ‘rJava’
* removing ‘/opt/conda/envs/R-3.6-py38/lib/R/library/rJava’

The downloaded source packages are in
	‘/tmp/Rtmp3aEbuj/downloaded_packages’
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Warning messages:
1: In doTryCatch(return(expr), name, parentenv, handler) :
  unable to load shared object '/opt/conda/envs/R-3.6-py38/lib/R/modules//R_X11.so':
  libXt.so.6: cannot open shared object file: No such file or directory
2: In install.packages("rJava") :
  installation of package ‘rJava’ had non-zero exit status
</pre></details>
I would love to look into the `config.log` file that is mentioned in the error message, but there is no such file anywhere. I've tried `R CMD INSTALL --no-clean-on-error`, but still no `config.log` anywhere.

Could you please give me some advice how to perform the installation, or how to get at the `console.log` file for further debugging?

**Comment by rolweber** [comment]
A potential problem I see is that I have the conda environment activated when calling `install.packages` or `R CMD INSTALL`, but that the installation procedure might use a script that runs without the conda environment activated.

If necessary, I'm OK with downloading the source package from CRAN and calling low-level installation commands instead of using `install.packages`. The feedstock for the [r-rjava conda package](https://github.com/conda-forge/r-rjava-feedstock/blob/master/recipe/build.sh#L6) uses `R CMD INSTALL --build . `, presumably after extracting the source archive.

**Comment by rolweber** [comment]
I'm currently trying to install on RedHat UBI 8 on x86_64. Eventually, I'll have to do something similar for ppc64le as well.

**Comment by rolweber** [comment]
Installation fails with `R CMD INSTALL --build .` too, but I get a `config.log` file. Within it:
```txt
configure:3852: checking whether JNI programs can be compiled
configure:3865: x86_64-conda_cos6-linux-gnu-cc -o conftest -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/R-3.6-py38/include -fdebug-prefix-map=/tmp/build/80754af9/r-base_1589917437985/work=/usr/local/src/conda/r-base-3.6.1 -fdebug-prefix-map=/opt/conda/envs/R-3.6-py38=/usr/local/src/conda-prefix -I/etc/alternatives/java_sdk_openjdk/include -I/etc/alternatives/java_sdk_openjdk/include/linux -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/R-3.6-py38/include -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/R-3.6-py38/lib -Wl,-rpath-link,/opt/conda/envs/R-3.6-py38/lib -L/opt/conda/envs/R-3.6-py38/lib conftest.c  -L/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server -ljvm >&5
/opt/conda/envs/R-3.6-py38/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server/libjvm.so: undefined reference to `memcpy@GLIBC_2.14'
/opt/conda/envs/R-3.6-py38/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /opt/conda/envs/R-3.6-py38/lib/libstdc++.so.6: undefined reference to `aligned_alloc@GLIBC_2.16'
/opt/conda/envs/R-3.6-py38/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /opt/conda/envs/R-3.6-py38/lib/libstdc++.so.6: undefined reference to `clock_gettime@GLIBC_2.17'
collect2: error: ld returned 1 exit status
configure:3865: $? = 1
```
Maybe a mismatch between the glibc versions in the conda environment and in the operating system.

**Comment by rolweber** [comment]
Several of the installed R packages require the conda gcc toolchain 7.x. I think the newer toolchain 9.3.0 would be a better match for the UBI 8 operating system and its OpenJDK.

I'll try to either bring the system libraries into a search path, or to symlink the system libraries into a conda directory that is on the search path. But it's possible that both fails.

```txt
x86_64-conda_cos6-linux-gnu-cc -o conftest -march=nocona -mtune=haswell -ftree-vectorize -fPIC
    -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe
    -isystem /opt/conda/envs/R-3.6-py38/include
    -fdebug-prefix-map=/tmp/build/80754af9/r-base_1589917437985/work=/usr/local/src/conda/r-base-3.6.1
    -fdebug-prefix-map=/opt/conda/envs/R-3.6-py38=/usr/local/src/conda-prefix
    -I/etc/alternatives/java_sdk_openjdk/include
    -I/etc/alternatives/java_sdk_openjdk/include/linux
    -DNDEBUG -D_FORTIFY_SOURCE=2 -O2
    -isystem /opt/conda/envs/R-3.6-py38/include
    -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections
    -Wl,-rpath,/opt/conda/envs/R-3.6-py38/lib
    -Wl,-rpath-link,/opt/conda/envs/R-3.6-py38/lib
    -L/opt/conda/envs/R-3.6-py38/lib
    conftest.c
    -L/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.302.b08-0.el8_4.x86_64/jre/lib/amd64/server
    -ljvm 
```

**Comment by rolweber** [comment]
`find / -name libc.so\*`
```txt
/usr/lib64/libc.so.6
/usr/lib64/libc.so
/opt/conda/envs/R-3.6-py38/x86_64-conda_cos6-linux-gnu/sysroot/lib/libc.so.6
/opt/conda/envs/R-3.6-py38/x86_64-conda_cos6-linux-gnu/sysroot/usr/lib/libc.so
/opt/conda/pkgs/gcc_impl_linux-64-7.3.0-habb00fd_1/x86_64-conda_cos6-linux-gnu/sysroot/lib/libc.so.6
/opt/conda/pkgs/gcc_impl_linux-64-7.3.0-habb00fd_1/x86_64-conda_cos6-linux-gnu/sysroot/usr/lib/libc.so
```

**Comment by rolweber** [comment]
`R CMD javareconf` compiles conftest.c into an object file and creates a shared library from that.
The configure script of the rJava package compiles and links conftest.c into an executable program.
It's that last step which doesn't work in my environment.

**Comment by rolweber** [comment]
I've given up on the original idea. The packages in Anaconda's R channel are built for the Anaconda compiler toolchain 7.x. The directory name `x86_64-conda_cos6-linux-gnu` refers to CentOS 6, which would match RedHat 6. I'm on RedHat UBI 8. The system libraries from the operating system don't mix well with those from the compiler toolchain.

The conda-forge channel has the latest rJava, but for compiler toolchain 9.3. That doesn't mix with the Anaconda R channel.

**Comment by rolweber** [comment]
I'm now using the following workaround to enable the OpenJDK from the operating system in the conda environment:

1. Build an empty conda package named "openjdk". It's just a placebo, no actual content required.
2. Install "r-rjava" from the Anaconda R channel, along with the placebo for "openjdk". This avoids pulling in the Anaconda-packaged OpenJDK. The R channel doesn't have the latest rJava, but the old version is good enough for my purposes.
3. Symlink `libjvm.so` from the RedHat OpenJDK into the conda environment as `$CONDA_PREFIX/lib/libjvm.so`.

   ```bash
   ln -sv $( find $JAVA_HOME/ -name libjvm.so -print -quit ) $CONDA_PREFIX/lib/
   if [[ ! -f $CONDA_PREFIX/lib/libjvm.so ]]; then
       >&2 echo "Error: symlinking the OpenJDK failed!"
       exit 1
   fi
   ```

**Comment by s-u** [comment]
@rolweber I'm not sure if it helps, but note that the JVM doesn't matter at all, since all JVMs have the same API. The actual location of the JVM is determined by R configuration at run-time, not by rJava nor its binary. That's why you're ok symlinking any JDK (BTW: you don't need a symlink if you configure R instead via `R CMD javareconf`).

I'm not sure I understand the exact details of the toolchain issues, but apparently whatever you using cannot create executables which looks like broken toolchain. rJava uses whatever compilers are configured by R so you just need to be consistent, it doesn't matter which chain it is as long as it is functioning and the same used by R.

**Comment by rolweber** [comment]
Thanks for the suggestions. I did `R CMD javareconf` before trying to install rJava. The problem is that there's a toolchain from Anaconda in the conda environment, which doesn't work with the system libraries from the operating system. I found no way to establishing consistency between the two, because the Anaconda R channel depends on a toolchain that matches CentOS/RedHat 6, while I'm installing on RedHat 8. The glibc versions differ too much.
The rJava package from Anaconda is a binary, which has been built in a consistent environment. By using that, I basically skip the installation steps that rJava performs for `R CMD INSTALL`.

Thanks for confirming that the libjvm.so API is stable. I'll try `R CMD javareconf` instead of the symlink next week.
